code/src/pages/product-details/components/product-details-container/ProductDetailsContainer.test.tsx

```diff
@@ -0,0 +1,167 @@
+>;
+
+const defaultArgs: MockState = {
+  data: null,
+  isLoading: false,
+  error: null
+};
+
+const productId = "1";
+
+const renderAndMock = (args?: MockState) => {
+  (useGetUserProductByIdQuery as jest.Mock).mockReturnValue({
+    ...defaultArgs,
+    ...args
+  });
+  renderWithProviders(<ProductDetailsContainer productId={productId} />);
+};
+
+describe("ProductDetailsContainer", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test("calls useGetUserProductByIdQuery with correct arguments", () => {
+    renderAndMock();
+
+    expect(useGetUserProductByIdQuery).toHaveBeenCalledWith({
+      productId,
+      lang: locale
+    });
+  });
+
+  test("renders loading element by default", () => {
+    renderAndMock({ isLoading: true });
+
+    const loadingElement = screen.getByText("Loading...");
+    expect(loadingElement).toBeInTheDocument();
+  });
+
+  test("redirects to 404 page when error with status 404 is returned from server", () => {
+    renderAndMock({ error: { status: 404 } });
+
+    expect(mockRenderRedirectComponent).toHaveBeenCalledWith(
+      productNotFoundRedirectConfig
+    );
+  });
+
+  test("does not render 404 page when isErrorWithStatus returns false", () => {
+    renderAndMock({ error: { message: "Test" } });
+
+    expect(mockRenderRedirectComponent).not.toHaveBeenCalled();
+  });
+
+  test("renders error element when there is an error but with status different from 404", () => {
+    renderAndMock({ error: { status: 500 } });
+
+    expect(mockRenderRedirectComponent).not.toHaveBeenCalled();
+
+    const errorElement = screen.getByText("Error...");
+    expect(errorElement).toBeInTheDocument();
+  });
+
+  test("renders error element when there is not error and no product", () => {
+    renderAndMock();
+
+    expect(mockRenderRedirectComponent).not.toHaveBeenCalled();
+
+    const errorElement = screen.getByText("Error...");
+    expect(errorElement).toBeInTheDocument();
+  });
+
+  test("renders product information correctly", () => {
+    renderAndMock({ data: mockProduct });
+
+    expect(mockRenderRedirectComponent).not.toHaveBeenCalled();
+
+    const productTitle = screen.getByText(mockProduct.name);
+    expect(productTitle).toBeInTheDocument();
+
+    const productPrice = screen.getByText(formatPrice(mockProduct.price));
+    expect(productPrice).toBeInTheDocument();
+
+    const productImage = screen.getByAltText(mockProduct.name);
+    expect(productImage).toHaveAttribute("src", mockProduct.image);
+
+    const categoryTag = screen.getByText("productsAll.mobile");
+    expect(categoryTag).toBeInTheDocument();
+
+    const inStockTypography = screen.getByText("productDetailsPage.inStock");
+    expect(inStockTypography).toBeInTheDocument();
+  });
+
+  test("renders delivery methods correctly", () => {
+    renderAndMock({ data: mockProduct });
+
+    const deliveryMethodImages = screen.getAllByAltText(
+      /dashboardTabs.orders.filters.\w+/
```

Comment:
will it work without the `regex` `\w+/` ? also can check other places with `regex`

Reply:
Let's consider following test:

```tsx
test("renders description correctly", () => {
    renderAndMock({
      data: { ...mockProduct, description: "value1/value2/value3/value4" }
    });

    const descriptionParagraphs = screen.getAllByText(/value\d/);
    expect(descriptionParagraphs).toHaveLength(4);
  });
```

We receive a string from backend that includes slashes that separate different characteristics of product. I just wanted to test that if we have description "value1/value2/value3/value4" then we should have 4 rendered paragraphs. I tried to use it without regex, but i got the error related to keys:

![image](https://github.com/user-attachments/assets/784da2aa-2d72-4c12-bb02-1805c4f876aa)


That's why i decided to use regex to match different items. I think it is quite readable. If you have a better suggestion then I am open to them

Let's also consider following test:

```
const deliveryMethodImages = screen.getAllByAltText(
      /dashboardTabs.orders.filters.\w+/
    );
```

I used regex here because for all delivery method images we have prefix `dashboardTabs.orders.filters.` and it will reduce code if we want to add new delivery method. I also modified assertion like follows:

```
expect(deliveryMethodImages).toHaveLength(deliveryMethods.length);
```

With this approach we won't have to change this test if we add new delivery method and it will pass